### PR TITLE
Add tests for `/make_knock` and `/send_knock` (same as the join counterparts) for when the federation server is in the middle of a partial join

### LINKED ISSUE #445: Add a test for `/make_knock` and `/send_knock` against a homeserver which is in the middle of a partial join.
Similar to #432 but for knocking.

Needs:
- [ ] https://github.com/matrix-org/gomatrixserverlib/pull/329
- [ ] https://github.com/matrix-org/gomatrixserverlib/pull/328
- [ ] TODO MakeKnock fed client function

### PR DESCRIPTION
**Pull request by reivilibre**
Fixes #445.

### CODE REVIEW

**Comment by DMRobertson** [comment]
Taking this out of the review queue since it's been re-marked as draft.

File: internal/federation/handle.go [hunk 1]
```diff
@@ -82,6 +82,37 @@ func MakeRespMakeJoin(s *Server, room *ServerRoom, userID string) (resp gomatrix
...
+		err = fmt.Errorf("make_knock cannot calculate auth_events: %w", err)
+		return
+	}
+	builder.AuthEvents = room.AuthEvents(stateNeeded)
+
+	resp = gomatrixserverlib.RespMakeKnock{
+		RoomVersion: room.Version,
+		KnockEvent:  builder,
+	}
+	return
```
Comment (DMRobertson) [inline internal/federation/handle.go:113]: Are we implicitly returning `resp, err` here?

File: tests/federation_room_join_partial_state_test.go [hunk 2]
```diff
@@ -1229,6 +1229,125 @@ func TestPartialStateJoin(t *testing.T) {
...
+		// We can't actually use /make_knock because host.docker.internal doesn't resolve,
+		// so compute it without making any requests:
+		makeKnockResp, err := federation.MakeRespMakeKnock(testServer1, serverRoom, testServer2.UserID("daniel"))
+		if err != nil {
+			t.Fatalf("MakeRespMakeKnock failed : %s", err)
+		}
+
+		// daniel then tries to /send_knock via the homeserver under test
+		knockEvent, err := makeKnockResp.KnockEvent.Build(time.Now(), gomatrixserverlib.ServerName(testServer2.ServerName()), testServer2.KeyID, testServer2.Priv, makeKnockResp.RoomVersion)
+		must.NotError(t, "KnockEvent.Build", err)
```
Comment (DMRobertson) [inline tests/federation_room_join_partial_state_test.go:1331]: is this just shorthand for `if err != nil { /* fail the test with given message */ }` ?

**Review by DMRobertson — APPROVED** [review]
SGTM. (I'm guessing that other than the MakeRespMakeKnock and GMSL changes that this was largely a find-and-replace for `join -> knock`?

File: tests/federation_room_join_partial_state_test.go (hunk 2 above)
Comment (reivilibre) [reply tests/federation_room_join_partial_state_test.go:1331]: yeah, seems to be
```go
// NotError will ensure `err` is nil else terminate the test with `msg`.
func NotError(t *testing.T, msg string, err error) {
	t.Helper()
	if err != nil {
		t.Fatalf("MustNotError: %s -> %s", msg, err)
	}
}
```

**Pull request merged**